apache/superset
# feat: apply Time Grain to X-Axis column

### SUMMARY

Currently, the `Time Grain` control defines a couple of `date truncate functions` on the `Time Column`. `Time Grain` only be applied if the `Time column` and `X-Axis` are set to the same value. This PR is the first step to removing the entire time section.

After this PR, when enabling the `GENERIC_CHART_AXES`

1. `Time Grain Control` removes from `Time Section`.
2. `Time Grain Control` will show if i) applies a Date Column in the `X-Axis`. ii) applies a Adhoc Column
3. `Time Grain Control` will be displayed under `X-Axis`.

**This change allows Time Column and Time Filter to use different columns**, Time Series Chart will support different column to set `Time Filter` and `X-Axis`, and `Time Grain` will be applied to the `X-Axis`.

#### Supported Charts
- Mixed Timeseries Chart
- Area  Chart
- Bar V2 Chart
- Line V2 Chart
- Scatter Chart
- SmoothLine Chart
- Step Chart

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
#### SQL Pattern

<img width="1287" alt="image" src="https://user-images.githubusercontent.com/2016594/187852878-e588793d-7ed9-4744-9322-9d5ebf571c0d.png">


### Apply `Categorical Column` or `Date Column` to the `X-Axis`
https://user-images.githubusercontent.com/2016594/187853599-abc7ff9c-957a-428d-b8ff-a6f9767d7a4a.mov



### Custom SQL

https://user-images.githubusercontent.com/2016594/187854928-4c566226-a8f6-4fca-9d85-bee7ba1798ec.mov




### TESTING INSTRUCTIONS

In order to test this PR, the multiple Date column should provide, so I prepare a CSV [(superstore_orders.csv)](https://github.com/apache/superset/files/9467696/superstore_orders.csv) file for that. Just load this CSV to test this PR.
1. Native filters should work as before
2. Time Grain should work as before
3. Time Grain should apply to X-Axis


### ADDITIONAL INFORMATION
